type.argument.inference.crashed for generic exception parameters since 3.43.0

### Description

Happens only with diamond operator. 
It's been broken since 3.43.0, last working version was 3.42.0.

### Command

```bash
javac \
-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED \
-J--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED \
-J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED \
-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED \
-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED \
-J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED \
-J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED \
-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED \
-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED \
-processorpath ./checker/dist/checker.jar \
-cp ./checker/dist/checker-qual.jar \
-processor org.checkerframework.checker.nullness.NullnessChecker \
-source 21 -target 21 \
./CheckerFailure.java
```

### File

```java
import java.io.IOException;

public final class CheckerFailure<E extends Exception> {

    public CheckerFailure(ThrowingFunctionalInterface<E> throwingFunctionalInterface) {}

    public static void main(String[] args) throws IOException {
        // Works
        new CheckerFailure<IOException>(() -> {
            throw new IOException("");
        });
        // Works
        CheckerFailure<IOException> b = new CheckerFailure<>(() -> {
            throw new IOException("");
        });
        // error: [type.argument.inference.crashed] Type argument inference crashed for CheckerFailure constructor
        new CheckerFailure<>(() -> {
            throw new IOException("");
        });
    }

    @FunctionalInterface
    public interface ThrowingFunctionalInterface<E extends Exception> {
        void throwingFunction() throws E;
    }

}
```

### Expected behaviour

Compilation succeeds.

### Actual behaviour

An exception is thrown:

```
CheckerFailure.java:17: error: [type.argument.inference.crashed] Type argument inference crashed for CheckerFailure constructor
        new CheckerFailure<>(() -> {
        ^
  error: An exception occurred: AnnotatedTypeFactory.getAnnotatedType: query of annotated type for tree THROW
  org.checkerframework.javacutil.BugInCF: AnnotatedTypeFactory.getAnnotatedType: query of annotated type for tree THROW
        at org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:1409)
        at org.checkerframework.framework.util.typeinference8.util.CheckedExceptionsUtil$CheckedExceptionATMVisitor.visitThrow(CheckedExceptionsUtil.java:219)
        at org.checkerframework.framework.util.typeinference8.util.CheckedExceptionsUtil$CheckedExceptionATMVisitor.visitThrow(CheckedExceptionsUtil.java:185)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCThrow.accept(JCTree.java:1761)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:92)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:111)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:272)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1104)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:92)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:96)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitLambdaExpression(TreeScanner.java:647)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1997)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:92)
        at org.checkerframework.framework.util.typeinference8.util.CheckedExceptionsUtil.thrownCheckedExceptionsATM(CheckedExceptionsUtil.java:178)
        at org.checkerframework.framework.util.typeinference8.types.InferenceFactory.getCheckedExceptionConstraints(InferenceFactory.java:997)
        at org.checkerframework.framework.util.typeinference8.constraint.CheckedExceptionConstraint.reduce(CheckedExceptionConstraint.java:70)
        at org.checkerframework.framework.util.typeinference8.constraint.ConstraintSet.reduceOneStep(ConstraintSet.java:329)
        at org.checkerframework.framework.util.typeinference8.constraint.ConstraintSet.reduceAdditionalArgOnce(ConstraintSet.java:309)
        at org.checkerframework.framework.util.typeinference8.InvocationTypeInference.getB4(InvocationTypeInference.java:671)
        at org.checkerframework.framework.util.typeinference8.InvocationTypeInference.infer(InvocationTypeInference.java:169)
        at org.checkerframework.framework.util.typeinference8.DefaultTypeArgumentInference.inferTypeArgs(DefaultTypeArgumentInference.java:110)
        at org.checkerframework.common.basetype.BaseTypeVisitor.checkTypeArgumentInference(BaseTypeVisitor.java:1883)
        ...
```

--- a/framework/src/main/java/org/checkerframework/framework/util/typeinference8/util/CheckedExceptionsUtil.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/typeinference8/util/CheckedExceptionsUtil.java
@@ -8,10 +8,12 @@ import com.sun.source.tree.ThrowTree;
 import com.sun.source.tree.TryTree;
 import com.sun.source.util.TreeScanner;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 import javax.lang.model.type.TypeKind;
 import javax.lang.model.type.TypeMirror;
 import javax.lang.model.type.UnionType;
+import org.checkerframework.checker.nullness.qual.Nullable;
 import org.checkerframework.framework.type.AnnotatedTypeMirror;
 import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
 import org.checkerframework.javacutil.TreeUtils;
@@ -31,14 +33,16 @@ public class CheckedExceptionsUtil {
    */
   public static List<TypeMirror> thrownCheckedExceptions(
       LambdaExpressionTree lambda, Java8InferenceContext context) {
-    return new CheckedExceptionVisitor(context).scan(lambda, null);
+    @Nullable List<TypeMirror> result = new CheckedExceptionVisitor(context).scan(lambda, null);
+    return result != null ? result : Collections.emptyList();
   }
 
   /**
    * Helper class for gathering the types of checked exceptions in a lambda. See
    * https://docs.oracle.com/javase/specs/jls/se9/html/jls-11.html#jls-11.2.2
    */
-  private static class CheckedExceptionVisitor extends TreeScanner<List<TypeMirror>, Void> {
+  private static class CheckedExceptionVisitor
+      extends TreeScanner<@Nullable List<TypeMirror>, Void> {
 
     /** the context. */
     private final Java8InferenceContext context;
@@ -70,7 +74,7 @@ public class CheckedExceptionsUtil {
       if (result == null) {
         result = new ArrayList<>();
       }
-      TypeMirror type = TreeUtils.typeOf(node);
+      TypeMirror type = TreeUtils.typeOf(node.getExpression());
       if (isCheckedException(type, context)) {
         result.add(type);
       }
@@ -171,7 +175,9 @@ public class CheckedExceptionsUtil {
    */
   public static List<AnnotatedTypeMirror> thrownCheckedExceptionsATM(
       LambdaExpressionTree lambda, Java8InferenceContext context) {
-    return new CheckedExceptionATMVisitor(context).scan(lambda, null);
+    @Nullable List<AnnotatedTypeMirror> result =
+        new CheckedExceptionATMVisitor(context).scan(lambda, null);
+    return result != null ? result : Collections.emptyList();
   }
 
   /**
@@ -179,7 +185,7 @@ public class CheckedExceptionsUtil {
    * https://docs.oracle.com/javase/specs/jls/se9/html/jls-11.html#jls-11.2.2
    */
   private static class CheckedExceptionATMVisitor
-      extends TreeScanner<List<AnnotatedTypeMirror>, Void> {
+      extends TreeScanner<@Nullable List<AnnotatedTypeMirror>, Void> {
 
     /** The context. */
     private final Java8InferenceContext context;
@@ -212,7 +218,7 @@ public class CheckedExceptionsUtil {
       if (result == null) {
         result = new ArrayList<>();
       }
-      AnnotatedTypeMirror type = context.typeFactory.getAnnotatedType(node);
+      AnnotatedTypeMirror type = context.typeFactory.getAnnotatedType(node.getExpression());
       if (isCheckedException(type, context)) {
         result.add(type);
       }

--- a/framework/tests/all-systems/java8inference/Issue6887.java
+++ b/framework/tests/all-systems/java8inference/Issue6887.java
@@ -1,0 +1,26 @@
+// https://github.com/typetools/checker-framework/issues/6887
+
+import java.io.IOException;
+
+public final class Issue6887<E extends Exception> {
+
+  public Issue6887(ThrowingFunctionalInterface<E> throwingFunctionalInterface) {}
+
+  public static void main(String[] args) throws IOException {
+    new Issue6887<IOException>(
+        () -> {
+          throw new IOException("");
+        });
+    new Issue6887<>(
+        () -> {
+          throw new IOException("");
+        });
+    new Issue6887<RuntimeException>(() -> {});
+    new Issue6887<>(() -> {});
+  }
+
+  @FunctionalInterface
+  public interface ThrowingFunctionalInterface<E extends Exception> {
+    void throwingFunction() throws E;
+  }
+}

